docs: Writing documentation for the built in core services (#15851)
backstage/backstage: docs/backend-system/core-services/01-index.md

### Using the service

The following example shows how to get a cache client in your `example` backend plugin and `set` and `get` values from the cache.

```ts
import {
  coreServices,
  createBackendPlugin,
} from '@backstage/backend-plugin-api';

createBackendPlugin({
  id: 'example',
  register(env) {
    env.registerInit({
      deps: {
        cache: coreServices.cache,
      },
      async init({ cache }) {
        const { key, value } = { key: 'test:key', value: 'bob' };
        await cache.set(key, value, { ttl: 1000 });

        // .. some other stuff.

        await cache.get(key); // 'bob'
      },
    });
  },
});
```

## Database

Interacting with databases inside your plugin is something that is quite common, and we provide a `PluginDatabaseManager` part of the core services that you can get `knex` client hooked up to your database which is configured in `app-config.yaml`.

If there's no config provided in `backend.database` then you will automatically get a simple in memory `sqlite3` client for your plugin.

These `PluginDatabaseManager`s are scoped per plugin too, so that table names do not conflict across plugins either.

### Using the service

> The following example shows how to get a `PluginDatabaseManager` in your `example` backend plugin and get a `client` for interacting with the database and running some migrations from a `migrationsDir` for your plugin.

Reviewer: ```suggestion
The following example shows how to get access to the database service in your `example` backend plugin and getting a client for interacting with the database. It also runs some migrations from a certain directory for your plugin.
```

```ts
import {
  coreServices,
  createBackendPlugin,
} from '@backstage/backend-plugin-api';

createBackendPlugin({
  id: 'example',
  register(env) {
    env.registerInit({
      deps: {
        database: coreServices.database,
      },
      async init({ database }) {

>         const client = database.getClient();

Reviewer: ```suggestion
        const client = await database.getClient();
```

        if (!database.migrations?.skip) {
          await client.migrate.latest({

>             directory: migrationsDir,

Reviewer: It's really nice that you show this. But I'm wondering if it's valuable to actually show the whole resolvePackagePath dance at the same time then, since it's a bit tricky to make certain you are getting the right path both locally and in prod. Eg https://github.com/backstage/backstage/blob/5b3948a4690fd168364a53c715401fcf6bb62c96/plugins/search-backend-module-pg/src/database/DatabaseDocumentStore.ts#L30

Maybe in a follow-up PR then, since that entails also showing where to put the migrations dir and adding it to the `"files"` section in the package.json as well ...

Could be an entire "Migrations" subsection actually, but again, in a later PR

          });
        }
      },
    });
  },
});
```

## Discovery

> When building plugins, you might find that you will need to lookup where in fact another plugins `baseUrl`. This could be for example, a `http` route or some `ws` protocol URL. For this we have the `discovery` service that you can query both the internal and external `baseUrl`s given a plugin ID.

Reviewer: ```suggestion
When building plugins, you might find that you will need to look up another plugin's base URL to be able to communicate with it. This could be for example an HTTP route or some `ws` protocol URL. For this we have a discovery service which can provide both internal and external base URLs for a given a plugin ID.
```

### Using the service

> The following example shows how to get the `DiscoveryService` in your `example` backend plugin and making a request to both the internal and external `baseUrl`s for the `derp` plugin.

Reviewer: ```suggestion
The following example shows how to get the discovery service in your `example` backend plugin and making a request to both the internal and external base URLs for the `derp` plugin.
```

```ts
import {
  coreServices,
  createBackendPlugin,
} from '@backstage/backend-plugin-api';

> import { fetch } from 'node-fetch';

Reviewer: No longer needed, right? It's native since node 15 and we're >16

createBackendPlugin({
  id: 'example',
  register(env) {
    env.registerInit({
      deps: {
        discovery: coreServices.discovery,
      },
      async init({ discovery }) {
        const urls = await Promise.all[
          discovery.getBaseUrl('derp'),
          discovery.getExternalBaseUrl('derp'),

>         ];

Reviewer: just make this into two separate await calls into two variables for clarity :) or a single getBaseUrl and a comment saying like "// Or use getExternalBaseUrl" if you ... blah blah

        await Promise.all(
          urls.map(
            (url) => fetch(url).then((r) => r.json()),
          ),
        );
      },
    });
  },
});
```

## Identity

> When working with backend plugins, you might find that you will need to interact with the `auth-backend` plugin to both authenticate backstage tokens, and get things like the `entityRef` of the authenticated user, and anything that they might claim to own through `ownershipEntityRefs`.

Reviewer: ```suggestion
When working with backend plugins, you might find that you will need to interact with the `auth-backend` plugin to both authenticate backstage tokens, and to deconstruct them to get the user's entity ref and/or ownership claims out of them.
```

### Using the service

> The following example shows how to get the `IdentityService` in your `example` backend plugin and retrieve the users `entityRef` and ownership claims for the incoming `http` request.

Reviewer: ```suggestion
The following example shows how to get the identity service in your `example` backend plugin and retrieve the user's entity ref and ownership claims for the incoming request.
```

```ts
import {
  coreServices,
  createBackendPlugin,
} from '@backstage/backend-plugin-api';
import { Router } from 'express';

createBackendPlugin({
  id: 'example',
  register(env) {
    env.registerInit({
      deps: {
        identity: coreServices.identity,
        http: coreServices.httpRouter,
      },
      async init({ http, identity }) {
        const router = Router();
        router.get('/test-me', (request, response) => {

>           // use the identityService pull out the header from the request and get the user

Reviewer: ```suggestion
          // use the identity service to pull out the header from the request and get the user
```

          const {
            identity: { userEntityRef, ownershipEntityRefs },
          } = await identity.getIdentity({
            request,
          });

>           // sent the decoded and validated things back to the user

Reviewer: ```suggestion
          // send the decoded and validated things back to the user
```

          response.json({
            userEntityRef,
            ownershipEntityRefs,
          });
        });

        http.use(router);
      },
    });
  },
});
```

### Configuration of the service

There's additional configuration that you can optionally pass to setup the `identity` core service.

> - `issuer` - Set an optional issuer for validation of the `jwt`

Reviewer: ```suggestion
- `issuer` - Set an optional issuer for validation of the JWT token
```

> - `algorithms` - `jws` `alg` header for validation of the `jwt`, defaults to `ES256`. More info on supported algorithms under [jose](https://github.com/panva/jose)

Reviewer: ```suggestion
- `algorithms` - `alg` header for validation of the JWT token, defaults to `ES256`. More info on supported algorithms can be found in the [`jose` library documentation](https://github.com/panva/jose)
```

You can configure these additional options by adding an override for the core service when calling `createBackend` like follows:

```ts

> import { identityFactory } from '@backstage/backend-app-api`;

Reviewer: ```suggestion
import { identityFactory } from '@backstage/backend-app-api';
```

const backend = createBackend({
  services: [
    identityFactory({
      issuer: 'backstage',
      algorithms: ['ES256', 'RS256']
    }),
  ],
});
```

## Lifecycle

> When writing plugins, it's often that you will have long running things that you might want to ensure clean shutdowns of when the plugins are torn down, or when the backend is quit (think local development). You shouldn't have to worry too much about providing shutdowns for any of the core services that you use, and should really only need to take care of anything that you create that you should stop when your plugin stops.

Reviewer: ```suggestion
This service allows your plugins to register hooks for cleaning up resources as the service is shutting down (e.g. when a pod is being torn down, or when pressing `Ctrl+C` during local development). Other core services also leverage this same mechanism internally to stop themselves cleanly.
```

Reviewer: Wasn't sure about the original core services formulation 🤔 Does this one make sense?

### Using the service

> The following example shows how to get the `LifecycleService` in your `example` backend plugin to clean a long running interval on teardown.

Reviewer: ```suggestion
The following example shows how to get the lifecycle service in your `example` backend plugin to clean up a long running interval when the service is shutting down.
```

```ts
import {
  coreServices,
  createBackendPlugin,
} from '@backstage/backend-plugin-api';

createBackendPlugin({
  id: 'example',
  register(env) {
    env.registerInit({
      deps: {
        lifecycle: coreServices.lifecycle,
        logger: coreServices.logger,
      },
      async init({ lifecycle, logger }) {

>         // setup by creating an interval that does something that we want to stop after the plugin is stopped.

Reviewer: ```suggestion
        // some example work that we want to stop when shutting down
```

        const interval = setInterval(async () => {
          await fetch('http://google.com/keepalive').then(r => r.json());
          // do some other stuff.

>         });

Reviewer: missing second parameter to setInterval ;)

        lifecycle.addShutdownHook({
          fn: () => clearInterval(interval),
          logger,
        });
      },
    });
  },
});
```

## Permissions

> Sometimes you want to include permissions and making sure that a user that is authorized to do some actions in your plugin. We've provide a core service out of the box for you to interact with the permissions framework. You can find out more about the permissions framework in [the documentation](https://backstage.io/docs/permissions/overview)

Reviewer: ```suggestion
This service allows your plugins to ask [the permissions framework](https://backstage.io/docs/permissions/overview) for authorization of user actions.
```

### Using the service

> The following example shows how to get the `PermissionsSerice` in your `example` backend to check to see if the user has the correct permissions for `myCustomPermission`.

Reviewer: ```suggestion
The following example shows how to get the permissions service in your `example` backend to check to see if the user is allowed to perform a certain action with a custom permission rule.
```

```ts
import {
  coreServices,
  createBackendPlugin,
} from '@backstage/backend-plugin-api';
import { Router } from 'express';

createBackendPlugin({
  id: 'example',
  register(env) {
    env.registerInit({
      deps: {
        permissions: coreServices.permissions,
        http: coreServices.httpRouter,
      },
      async init({ permissions, http }) {
        const router = Router();
        router.get('/test-me', (request, response) => {

>           // use the identityService pull out the header from the request and get the token

Reviewer: ```suggestion
          // use the identity service to pull out the token from request headers
```

          const { token } = await identity.getIdentity({
            request,
          });

          // ask the permissions framework what the decision is for the permission
          const permissionResponse = await permissions.authorize(
            [
              {
                permission: myCustomPermission,
              },
            ],
            { token },
          );
        });

        http.use(router);
      },
    });
  },
});
```

## Scheduler

> When writing plugins, it's often that you want to have things running on a schedule, or something similar to cron jobs that are distributed through instances that your backend plugin might be running on. We supply a `TaskScheduler` that is scoped per plugin so that you can create these tasks and orchestrate the running of them.

Reviewer: ```suggestion
When writing plugins, you sometimes want to have things running on a schedule, or something similar to cron jobs that are distributed through instances that your backend plugin is running on. We supply a task scheduler for this purpose that is scoped per plugin so that you can create these tasks and orchestrate their execution.
```

### Using the service

> The following example shows how to get the `SchedulerService` in your `example` backend to schedule a scheduled task that runs once across your instances at a given interval.

Reviewer: ```suggestion
The following example shows how to get the scheduler service in your `example` backend to issue a scheduled task that runs across your instances at a given interval.
```

```ts
import {
  coreServices,
  createBackendPlugin,
} from '@backstage/backend-plugin-api';

> import { fetch } from 'node-fetch';

Reviewer: remove

Author: Answered in https://github.com/backstage/backstage/pull/15851#discussion_r1082442134

createBackendPlugin({
  id: 'example',
  register(env) {
    env.registerInit({
      deps: {
        scheduler: coreServices.scheduler,
      },
      async init({ scheduler }) {
        await scheduler.scheduleTask({

>           frequency: Duration.fromObject({ minutes: 10 }),

Reviewer: ```suggestion
          frequency: { minutes: 10 },
          timeout: { seconds: 30 },
```

love the HumanDuration shorthand so we don't need to pull in luxon.

and the timeout is mandatory actually

### Using the service

The following example shows how to get the logger in your `example` backend plugin and create a warning message that will be printed nicely to the console.

```ts
import {
  coreServices,
  createBackendPlugin,
} from '@backstage/backend-plugin-api';

createBackendPlugin({
  id: 'example',
  register(env) {
    env.registerInit({
      deps: {
        log: coreServices.logger,
      },
      async init({ log }) {
        log.warn("Here's a nice log line that's a warning!");
      },
    });
  },
});
```

### Root Logger

The root logger is the logger that is used by other root services. It's where the implemenation lies for creating child loggers around the backstage ecosystem including child loggers for plugins with the correct metadata and annotations.

If you want to override the implementation for logging across all of the backend, this is the service that you should override.

### Configuring the service

The following example is how you can override the root logger service to add additional metadata to all log lines.

```ts

> import { coreServices } from '@backstage/backend-plugin-api';

Reviewer: maybe add winston imports, but no biggie

### Using the service

The following example shows how to get the lifecycle service in your `example` backend plugin to clean up a long running interval when the service is shutting down.

```ts
import {
  coreServices,
  createBackendPlugin,
} from '@backstage/backend-plugin-api';

createBackendPlugin({
  id: 'example',
  register(env) {
    env.registerInit({
      deps: {
        lifecycle: coreServices.lifecycle,
        logger: coreServices.logger,
      },
      async init({ lifecycle, logger }) {
        // some example work that we want to stop when shutting down
        const interval = setInterval(async () => {
          await fetch('http://google.com/keepalive').then(r => r.json());
          // do some other stuff.
        }, 1000);

        lifecycle.addShutdownHook({
          fn: () => clearInterval(interval),
          logger,
        });
      },
    });
  },
});
```

## Root Lifecycle

> This service is the same as the lifecycle service, but should only be used by the root services. This is also where the implementation for the actual lifecycle hooks torn, so if you want to override the implementation of how the lifecycle hooks are executed, you should override this service.

Reviewer: torn? 🤔

Author: oh what where did that come from

        await scheduler.scheduleTask({
          frequency: { minutes: 10 },
          timeout: { seconds: 30 },
          id: 'ping-google',
          fn: async () => {
            await fetch('http://google.com/ping');
          },
        });
      },
    });
  },
});
```

## URL Readers

Plugins will require communication with certain integrations that users have configured. Popular integrations are things like Version Control Systems (VSC), such as GitHub, BitBucket GitLab etc. These integrations are configured in the `integrations` section of the `app-config.yaml` file.

These URL readers are basically wrappers with authentication for files and folders that could be stored in these VCS repositories.

### Using the service

The following example shows how to get the URL Reader service in your `example` backend plugin to read a file and a directory from a GitHub repository.

```ts
import {
  coreServices,
  createBackendPlugin,
} from '@backstage/backend-plugin-api';
import os from 'os';

createBackendPlugin({
  id: 'example',
  register(env) {
    env.registerInit({
      deps: {
        urlReader: coreServices.urlReader,
      },
      async init({ urlReader }) {

>         const reader = await urlReader

Reviewer: ```suggestion
        const buffer = await urlReader
```
